transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala

```diff
@@ -99,24 +113,78 @@ object HcaPipelineBuilder extends PipelineBuilder[Args] {
     * @return a Msg object in the desired output format
     */
   def transformMetadata(entityType: String, fileName: String, metadata: Msg): Msg = {
-    // extract info from file name
+    val (entityId, entityVersion) = getEntityIdAndVersion(fileName: String)
+    Obj(
+      mutable.LinkedHashMap[Msg, Msg](
+        Str(s"${entityType}_id") -> Str(entityId),
+        Str("version") -> Str(entityVersion),
+        Str("content") -> Str(encode(metadata).getOrElse(""))
+      )
+    )
+  }
+
+  /**
+    *
+    * Extract the necessary info from a file of file-related metadata and put it into a form that
+    * makes it easy to pass in to the table format
+    *
+    * @param entityType the metadata entity type to prepend for the id field
+    * @param fileName the raw filename of the metadata file
+    * @param metadata the content of the metadata file in Msg format
+    * @return a Msg object in the desired output format
+    */
+  def transformFileMetadata(entityType: String, fileName: String, metadata: Msg): Msg = {
+    val (entityId, entityVersion) = getEntityIdAndVersion(fileName: String)
+    val dataFileName = metadata.read[Msg]("file_core").read[String]("file_name")
+    val (fileId, fileVersion) = getFileIdAndVersion(dataFileName)
+    // put values in the form we want
+    Obj(
+      mutable.LinkedHashMap[Msg, Msg](
+        Str(s"${entityType}_id") -> Str(entityId),
+        Str("version") -> Str(entityVersion),
+        Str("content") -> Str(encode(metadata).getOrElse("")),
+        Str("source_file_id") -> Str(fileId),
+        Str("source_file_version") -> Str(fileVersion)
+      )
+    )
+  }
+
+  /**
+    * Extract the entity id and entity version from the name of a metadata file.
+    *
+    * @param fileName the raw filename of the metadata file
+    * @return a tuple of the entity id and entity version
+    */
+  def getEntityIdAndVersion(fileName: String): (String, String) = {
     val matches = metadataPattern
       .findFirstMatchIn(fileName)
       .getOrElse(
         throw new Exception(
-          s"transformMetadata: error when finding id and version from file named $fileName"
+          s"transformMetadata: error when finding entity id and version from file named $fileName"
         )
       )
     val entityId = matches.group(1)
     val entityVersion = matches.group(2)
-    // and put in form we want
-    Obj(
-      mutable.LinkedHashMap[Msg, Msg](
-        Str(s"${entityType}_id") -> Str(entityId),
-        Str("version") -> Str(entityVersion),
-        Str("content") -> Str(encode(metadata).getOrElse(""))
+    (entityId, entityVersion)
+  }
+
+  /**
+    * Extract the file id and file version from the name of a data file.
+    *
+    * @param fileName the raw filename of the data file
+    * @return a tuple of the file id and file version
+    */
+  def getFileIdAndVersion(fileName: String): (String, String) = {
```

Comment:
This is good for now, but I expect we'll want to also return the directory path as part of the follow-on ticket